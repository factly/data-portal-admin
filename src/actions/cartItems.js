import axios from '../utils/axios';
import {
  ADD_CARTITEMS,
  SET_CARTITEM_LOADING,
  SET_CARTITEM_REQUEST,
  SET_CARTITEM_IDS,
  CARTITEM_API,
} from '../constants/cartItems';
import { addProducts } from './products';
import { getIds, getValues, deleteKeys, buildObjectOfItems } from '../utils/objects';

export const loadCartItems = (id, page = 1, limit = 5) => {
  return async (dispatch, getState) => {
    const {
      cartItems: { req },
    } = getState();

    let ids;
    for (let item of req) {
      if (item.page === page && item.limit === limit) {
        ids = [...item.ids];
      }
    }

    if (ids) {
      dispatch(setCartItemIds(ids));
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${CARTITEM_API(id)}?page=${page}&limit=${limit}`,
      method: 'get',
    });

    const { nodes, total } = response.data;
    const currentPageIds = getIds(nodes);
    const currentReq = { id, page: page, limit: limit, ids: currentPageIds };
    dispatch(setCartItemRequest(currentReq, total));
    dispatch(addCartItems(nodes));
    dispatch(setCartItemIds(currentPageIds));

    dispatch(setLoading(false));
  };
};

const setLoading = (loading) => {
  return {
    type: SET_CARTITEM_LOADING,
    payload: { loading },
  };
};

export const addCartItems = (cartItems) => (dispatch) => {
  const products = getValues(cartItems, 'product');
  dispatch(addProducts(products));

  dispatch({
    type: ADD_CARTITEMS,
    payload: {
      cartItems: buildObjectOfItems(deleteKeys(cartItems, ['product'])),
    },
  });
};

const setCartItemRequest = (req, total) => {
  return {
    type: SET_CARTITEM_REQUEST,
    payload: { req, total },
  };
};

const setCartItemIds = (ids) => {
  return {
    type: SET_CARTITEM_IDS,
    payload: { ids },
  };
};
