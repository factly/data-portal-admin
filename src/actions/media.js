import axios from '../utils/axios';
import {
  ADD_MEDIUM,
  ADD_MEDIA,
  SET_MEDIUM_LOADING,
  SET_MEDIUM_REQUEST,
  SET_MEDIUM_IDS,
  RESET_MEDIUM,
  MEDIUM_API,
} from '../constants/media';
import { getIds, buildObjectOfItems } from '../utils/objects';

export const loadMedia = (page = 1, limit = 5) => {
  return async (dispatch, getState) => {
    dispatch(setLoading(true));

    const response = await axios({
      url: `${MEDIUM_API}?page=${page}&limit=${limit}`,
      method: 'get',
    });

    const { nodes, total } = response.data;
    const currentPageIds = getIds(nodes);
    const currentReq = { page: page, limit: limit, ids: currentPageIds };
    dispatch(setMediumRequest(currentReq, total));
    dispatch(addMedia(nodes));
    dispatch(setMediumIds(currentPageIds));

    dispatch(setLoading(false));
  };
};

export const createMedium = (data) => {
  return async (dispatch, getState) => {
    dispatch(setLoading(true));

    await axios({
      url: MEDIUM_API,
      method: 'post',
      data: data,
    });

    dispatch(resetMedium());
    dispatch(setLoading(false));
  };
};

export const updateMedium = (id, data) => {
  return async (dispatch, getState) => {
    let url = `${MEDIUM_API}/${id}`;

    dispatch(setLoading(true));

    const response = await axios({
      url: url,
      method: 'put',
      data: data,
    });

    dispatch(addMedium(response.data));
    dispatch(setLoading(false));
  };
};

export const deleteMedium = (id) => {
  return async (dispatch, getState) => {
    let url = `${MEDIUM_API}/${id}`;

    dispatch(setLoading(true));

    await axios({
      url: url,
      method: 'delete',
    });

    dispatch(resetMedium());
    dispatch(setLoading(false));
  };
};

export const getMedium = (id) => {
  return async (dispatch, getState) => {
    const {
      media: { items },
    } = getState();

    if (items[id]) {
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${MEDIUM_API}/${id}`,
      method: 'get',
    });
    dispatch(addMedium(response.data));

    dispatch(setLoading(false));
  };
};

export const setLoading = (loading) => {
  return {
    type: SET_MEDIUM_LOADING,
    payload: { loading },
  };
};

export const addMedium = (medium) => {
  return {
    type: ADD_MEDIUM,
    payload: { medium },
  };
};

export const addMedia = (media) => {
  return {
    type: ADD_MEDIA,
    payload: {
      media: buildObjectOfItems(media),
    },
  };
};

export const setMediumRequest = (req, total) => {
  return {
    type: SET_MEDIUM_REQUEST,
    payload: { req, total },
  };
};

export const setMediumIds = (ids) => {
  return {
    type: SET_MEDIUM_IDS,
    payload: { ids },
  };
};

export const resetMedium = () => {
  return {
    type: RESET_MEDIUM,
  };
};
