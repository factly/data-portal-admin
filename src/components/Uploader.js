import React from 'react';
import Uppy from '@uppy/core';
import AwsS3 from '@uppy/aws-s3';
import GoogleDrive from '@uppy/google-drive';
import Url from '@uppy/url';
import { DashboardModal } from '@uppy/react';
import '@uppy/core/dist/style.css';
import '@uppy/dashboard/dist/style.css';
import '@uppy/url/dist/style.css';

import { Button } from 'antd';

const Uploader = ({ onUploadSuccess, metaFields }) => {
  const [show, setShow] = React.useState(false);

  const uppy = Uppy({
    id: 'uppy-media',
    meta: { type: 'avatar' },
    restrictions: {
      maxNumberOfFiles: 1,
    },
    autoProceed: false,
    onBeforeUpload: (files) => {
      const updatedFiles = {};
      Object.keys(files)
        .filter((fileId) => files[fileId].extension)
        .forEach((fileId) => {
          updatedFiles[fileId] = files[fileId];
        });
      if (Object.keys(updatedFiles).length === 0) {
        uppy.info('File must have a valid file extension');
        return false;
      }

      return updatedFiles;
    },
  })
    .use(AwsS3, { companionUrl: 'http://localhost:3020' })
    .use(Url, { companionUrl: 'http://localhost:3020' })
    .use(GoogleDrive, { companionUrl: 'http://localhost:3020' })
    .on('complete', (result) => {
      onUploadSuccess(result.successful);
      setShow(false);
    });

  return (
    <div>
      <Button onClick={() => setShow(true)}>Upload</Button>
      <DashboardModal
        uppy={uppy}
        closeModalOnClickOutside
        open={show}
        onRequestClose={() => setShow(false)}
        plugins={['GoogleDrive', 'Url']}
        metaFields={metaFields}
      />
    </div>
  );
};

export default Uploader;
