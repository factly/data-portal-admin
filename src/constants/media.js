export const ADD_MEDIUM = 'ADD_MEDIUM';
export const ADD_MEDIA = 'ADD_MEDIA';
export const SET_MEDIUM_LOADING = 'SET_MEDIUM_LOADING';
export const SET_MEDIUM_REQUEST = 'SET_MEDIUM_REQUEST';
export const SET_MEDIUM_IDS = 'SET_MEDIUM_IDS';
export const RESET_MEDIUM = 'RESET_MEDIUM';

export const MEDIUM_API = `/media`;
