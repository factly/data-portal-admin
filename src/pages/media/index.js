import React from 'react';

import MediaList from './components/MediaList';
import MediaUploader from './components/MediaUpload';

function Media() {
  return (
    <>
      <MediaUploader />
      <MediaList />
    </>
  );
}

export default Media;
