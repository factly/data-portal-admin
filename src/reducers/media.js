import {
  ADD_MEDIA_LIST_REQUEST,
  SET_MEDIA_LIST_CURRENT_PAGE,
  LOADING_MEDIA,
  LOAD_MEDIA_SUCCESS,
  SET_MEDIA_LIST_TOTAL,
  LOAD_MEDIA_FAILURE,
  GET_MEDIUM_SUCCESS,
  GET_MEDIUM_FAILURE,
  CREATING_MEDIUM,
  CREATE_MEDIUM_SUCCESS,
  CREATE_MEDIUM_FAILURE,
  UPDATE_MEDIUM_SUCCESS,
  UPDATE_MEDIUM_FAILURE,
  DELETE_MEDIUM_SUCCESS,
  DELETE_MEDIUM_FAILURE,
} from '../constants/media';

const initialState = {
  loading: false,
  ids: [],
  req: [],
  items: {},
  details: {},
  medium: {},
  total: 0,
};

export default function mediaReducer(state = initialState, action = {}) {
  switch (action.type) {
    case LOADING_MEDIA:
      return {
        ...state,
        loading: true,
      };
    case LOAD_MEDIA_SUCCESS: {
      const { items } = action.payload;
      return {
        ...state,
        loading: false,
        items: { ...state.items, ...items },
      };
    }
    case SET_MEDIA_LIST_TOTAL:
      return {
        ...state,
        total: action.payload,
      };
    case ADD_MEDIA_LIST_REQUEST: {
      return {
        ...state,
        req: [...state.req, action.payload],
      };
    }
    case SET_MEDIA_LIST_CURRENT_PAGE:
      return {
        ...state,
        ids: action.payload,
      };
    case LOAD_MEDIA_FAILURE:
      return {
        ...state,
        loading: false,
      };
    case GET_MEDIUM_SUCCESS:
      return {
        ...state,
        medium: action.payload,
      };
    case GET_MEDIUM_FAILURE:
      return {
        ...state,
        loading: false,
      };
    case CREATING_MEDIUM:
      return {
        ...state,
        loading: true,
      };
    case CREATE_MEDIUM_SUCCESS: {
      const medium = action.payload;
      return {
        ...state,
        loading: false,
        req: [],
        items: { ...state.items, [medium.id]: medium },
        total: state.total + 1,
      };
    }
    case CREATE_MEDIUM_FAILURE:
      return {
        ...state,
        loading: false,
      };
    case UPDATE_MEDIUM_SUCCESS: {
      const medium = action.payload;
      return {
        ...state,
        loading: false,
        medium: {},
        items: { ...state.items, [medium.id]: medium },
      };
    }
    case UPDATE_MEDIUM_FAILURE:
      return {
        ...state,
        loading: false,
      };
    case DELETE_MEDIUM_SUCCESS: {
      const id = action.payload;
      const newItems = { ...state.items };
      delete newItems[id];
      return {
        ...state,
        loading: false,
        req: [],
        ids: [],
        items: newItems,
        total: state.total - 1,
      };
    }
    case DELETE_MEDIUM_FAILURE:
      return {
        ...state,
        loading: false,
      };
    default:
      return state;
  }
}
