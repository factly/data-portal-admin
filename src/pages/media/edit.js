import React from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { Row, Col, Skeleton, Form, Input, Button, notification } from 'antd';
import { useSelector, useDispatch } from 'react-redux';
import { getMedium, updateMedium } from '../../actions/media';

const layout = {
  labelCol: {
    span: 8,
  },
  wrapperCol: {
    span: 16,
  },
};

const tailLayout = {
  wrapperCol: { offset: 8, span: 16 },
};

function MediaEdit() {
  const [form] = Form.useForm();
  const history = useHistory();
  const { id } = useParams();
  const dispatch = useDispatch();
  const { media, loading } = useSelector(({ media }) => {
    return {
      media: media.items[id],
      loading: media.loading,
    };
  });

  React.useEffect(() => {
    dispatch(getMedium(id));
  }, [id]);

  const updateMedia = (values) => {
    const data = {
      ...media,
      ...values,
    };
    dispatch(updateMedium(id, data))
      .then(() => {
        notification.success({
          message: 'Success',
        });
        history.push('/media');
      })
      .catch((error) => {
        console.log(error);
        console.notification.error({
          message: 'Error',
          description: 'Something went wrong',
        });
      });
  };

  if (loading) return <Skeleton />;
  return (
    <Row>
      <Col span={'12'}>
        <img src={media.url} alt={'space'} style={{ width: '100%' }} />
      </Col>
      <Col span={'12'}>
        <Form
          {...layout}
          form={form}
          name="create-space"
          onFinish={(values) => {
            updateMedia(values);
          }}
          initialValues={media}
        >
          <Form.Item name="name" label="Name">
            <Input />
          </Form.Item>
          <Form.Item name="alt_text" label="Alt Text">
            <Input />
          </Form.Item>
          <Form.Item name="caption" label="Caption">
            <Input />
          </Form.Item>
          <Form.Item name="description" label="Description">
            <Input />
          </Form.Item>
          <Form.Item {...tailLayout}>
            <Button type="primary" htmlType="submit">
              Submit
            </Button>
          </Form.Item>
        </Form>
      </Col>
    </Row>
  );
}

export default MediaEdit;
