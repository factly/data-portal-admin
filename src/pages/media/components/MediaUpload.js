import React from 'react';
import { useDispatch } from 'react-redux';
import { notification } from 'antd';

import { createMedium } from '../../../actions/media';
import Uploader from '../../../components/Uploader';

function MediaUploader({ onUploadSuccess = () => {} }) {
  const dispatch = useDispatch();

  const onFinish = (values) => {
    try {
      Object.values(values).map((file) => {
        const data = {};
        data['alt_text'] = file.meta.caption;
        data['caption'] = file.meta.caption;
        data['description'] = file.meta.description;
        data['dimensions'] = '100x100';
        data['file_size'] = file.size;
        data['name'] = file.meta.name;
        data['slug'] = file.meta.caption;
        data['title'] = file.meta.caption;
        data['type'] = file.meta.type;
        data['url'] = file.uploadURL;

        dispatch(createMedium(data))
          .then((data) => {
            notification.success({
              message: 'Success',
            });
            onUploadSuccess(data);
          })
          .catch((error) => {
            notification.error({
              message: 'Error',
              description: 'Something went wrong',
            });
          });
      });
    } catch (error) {
      console.log(error);
    }
  };

  const metaFields = [
    { id: 'name', name: 'Name', placeholder: 'file name' },
    { id: 'caption', name: 'Caption', placeholder: 'add a caption' },
    {
      id: 'description',
      name: 'Description',
      placeholder: 'describe what the image is about',
    },
  ];
  return (
    <div>
      <Uploader onUploadSuccess={onFinish} metaFields={metaFields} />
    </div>
  );
}

export default MediaUploader;
