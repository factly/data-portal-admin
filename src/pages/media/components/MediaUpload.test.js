import React from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch, Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount, shallow } from 'enzyme';
import { act } from '@testing-library/react';

import '../../../matchMedia.mock';
import MediaUpload from './MediaUpload';
import * as actions from '../../../actions/media';
import Uploader from '../../../components/Uploader';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));
jest.mock('../../../actions/media', () => ({
  createMedium: jest.fn(),
}));

describe('Media upload component', () => {
  let store;
  let mockedDispatch;
  store = mockStore({});
  store.dispatch = jest.fn(() => ({}));
  mockedDispatch = jest.fn(() => Promise.resolve({}));
  useDispatch.mockReturnValue(mockedDispatch);

  describe('snapshot testing', () => {
    it('should render the component', () => {
      const component = shallow(<MediaUpload />);
      expect(component).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    afterEach(() => {
      wrapper.unmount();
    });
    it('should call createMedium', (done) => {
      actions.createMedium.mockReset();
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <MediaUpload />
          </Provider>,
        );
      });
      wrapper.find('Button').simulate('click');
      wrapper
        .find('Uploader')
        .props()
        .onUploadSuccess([
          {
            meta: {
              caption: 'caption',
              description: 'description',
              name: 'name',
              type: 'jpeg',
            },
            uploadURL: 'url',
            size: 100,
          },
        ]);
      setTimeout(() => {
        expect(actions.createMedium).toHaveBeenCalledWith({
          caption: 'caption',
          description: 'description',
          name: 'name',
          type: 'jpeg',
          slug: 'caption',
          title: 'caption',
          alt_text: 'caption',
          url: 'url',
          dimensions: '100x100',
          file_size: 100,
        });
        done();
      }, 0);
    });
  });
});
