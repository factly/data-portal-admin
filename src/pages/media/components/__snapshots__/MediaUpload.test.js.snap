// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Media upload component snapshot testing should render the component 1`] = `
<div>
  <Uploader
    metaFields={
      Array [
        Object {
          "id": "name",
          "name": "Name",
          "placeholder": "file name",
        },
        Object {
          "id": "caption",
          "name": "Caption",
          "placeholder": "add a caption",
        },
        Object {
          "id": "description",
          "name": "Description",
          "placeholder": "describe what the image is about",
        },
      ]
    }
    onUploadSuccess={[Function]}
  />
</div>
`;
